fw/http_match.c

```diff
@@ -661,40 +732,91 @@ tfw_http_arg_adjust(const char *arg, tfw_http_match_fld_t field,
 	 * pattern should be applied.
 	 */
 	if (!wc_arg && arg[0] == '*') {
-		if (*op_out == TFW_HTTP_MATCH_O_PREFIX)
+		if (*op_out == TFW_HTTP_MATCH_O_PREFIX) {
 			T_WARN_NL("http_match: unable to match"
 				  " double-wildcard patterns '%s', so"
 				  " prefix pattern will be applied\n", arg);
-
-		else if (raw_hdr_name)
-			T_WARN_NL("http_match: unable to match suffix"
-				  " pattern '%s' in case of raw header"
-				  " specification: '%s', so wildcard pattern"
-				  " will not be applied\n", arg, raw_hdr_name);
-
-		else
+		}
+		else if (raw_hdr_name) {
+			if (field != TFW_HTTP_MATCH_F_COOKIE)
+				T_WARN_NL("http_match: unable to match suffix"
+					  " pattern '%s' in case of raw header"
+					  " specification: '%s', so wildcard"
+					  " pattern will not be applied\n",
+					  arg, raw_hdr_name);
+			else
+				*op_out = TFW_HTTP_MATCH_O_SUFFIX;
+		} else {
 			*op_out = TFW_HTTP_MATCH_O_SUFFIX;
+		}
 	}
 
-	len = full_name_len;
-	escaped = false;
 	pos = arg_out + full_name_len;
-	for (i = 0; arg[i]; ++i) {
-		if (arg[i] == '*' && !escaped && (i == 0 || !arg[i + 1]))
-			continue;
-		if (arg[i] != '\\' || escaped) {
-			escaped = false;
-			*pos = arg[i];
-			++len;
-			++pos;
-		}
-		else if (arg[i] == '\\') {
-			escaped = true;
+	len = tfw_http_escape_pre_post(pos, arg);
+	*size_out += full_name_len + len + 1;
+
+	return arg_out;
+}
+
+const char *
+tfw_http_val_adjust(const char *val, tfw_http_match_fld_t field,
+		    unsigned int *len_out,
+		    tfw_http_match_val_t *type_out,
+		    tfw_http_match_op_t *op_out)
+{
+	size_t len, len_adjust;
+	char *val_out;
+	bool wc_val;
+
+	if (field == TFW_HTTP_MATCH_F_HDR) {
+		*type_out = TFW_HTTP_MATCH_V_HID;
+		return NULL;
+	}
+	else if (field == TFW_HTTP_MATCH_F_COOKIE)
+	{
+		*type_out = TFW_HTTP_MATCH_V_COOKIE;
+	} else {
+		/* When not a hdr or cookie rule this value is not used */
+		return NULL;
+	}
+
+	if (!val) {
+		T_ERR_NL("http_tbl: cookie pattern is empty, must be filled\n");
+		return ERR_PTR(-EINVAL);
+	}
+
+	len = strlen(val);
+	wc_val = (val[0] == '*' && len == 1);
+
+	if (!(val_out = kzalloc(len + SLEN("=") + 1, GFP_KERNEL))) {
```

Comment:
Well, the dynamic allocation isn't good since we pay with worse memory locality. However, this isn't the only place where we have the allocation. I updated #732 with the statement about memory. In general, we should rework the rules engine. But there is nothing to do for now.

Reply:
Yes, that is reasonable.